Mathlib/RingTheory/DividedPowers/Basic.lean

```diff
@@ -0,0 +1,514 @@
+def dividedPowersBot (A : Type*) [CommSemiring A] [DecidableEq A] :
+    DividedPowers (⊥ : Ideal A) where
+  dpow n a := ite (a = 0 ∧ n = 0) 1 0
+  dpow_null {n a} ha := by
+    simp only [Ideal.mem_bot] at ha
+    dsimp
+    rw [if_neg]
+    exact not_and_of_not_left (n = 0) ha
+  dpow_zero {a} ha := by
+    rw [Ideal.mem_bot.mp ha]
+    simp only [and_self, ite_true]
+  dpow_one {a} ha := by
+    simp [Ideal.mem_bot.mp ha]
+  dpow_mem {n a} hn _ := by
+    simp only [Ideal.mem_bot, ite_eq_right_iff, and_imp]
+    exact fun _ a => False.elim (hn a)
+  dpow_add n a b ha hb := by
+    rw [Ideal.mem_bot.mp ha, Ideal.mem_bot.mp hb, add_zero]
+    simp only [true_and, ge_iff_le, tsub_eq_zero_iff_le, mul_ite, mul_one, mul_zero]
+    split_ifs with h
+    · simp [h]
+    · symm
+      apply sum_eq_zero
+      intro i hi
+      simp only [mem_antidiagonal] at hi
+      split_ifs with h2 h1
+      · rw [h1, h2, add_zero] at hi
+        exfalso
+        exact h hi.symm
+      · rfl
+      · rfl
+  dpow_smul n {a x} hx := by
+    rw [Ideal.mem_bot.mp hx]
+    simp only [mul_zero, true_and, mul_ite, mul_one]
+    by_cases hn : n = 0
+    · rw [if_pos hn, hn, if_pos rfl, _root_.pow_zero]
+    · simp only [if_neg hn]
+  dpow_mul m n {x} hx := by
+    rw [Ideal.mem_bot.mp hx]
+    simp only [true_and, mul_ite, mul_one, mul_zero, add_eq_zero]
+    by_cases hn : n = 0
+    · simp only [hn, ite_true, and_true, add_zero, choose_self, cast_one]
+    · rw [if_neg hn, if_neg]
+      exact not_and_of_not_right (m = 0) hn
+  dpow_comp m {n a} hn ha := by
+    rw [Ideal.mem_bot.mp ha]
+    simp only [true_and, ite_eq_right_iff, _root_.mul_eq_zero, mul_ite, mul_one, mul_zero]
+    by_cases hm: m = 0
+    · simp only [hm, and_true, true_or, ite_true, uniformBell_zero_left, cast_one]
+      rw [if_pos]
+      exact fun h => False.elim (hn h)
+    · simp only [hm, and_false, ite_false, false_or, if_neg hn]
+
+instance {A : Type*} [CommSemiring A] [DecidableEq A] :
+    Inhabited (DividedPowers (⊥ : Ideal A)) := ⟨dividedPowersBot A⟩
+
+/-- The coercion from the divided powers structures to functions -/
+instance {A : Type*} [CommSemiring A] (I : Ideal A) :
+    CoeFun (DividedPowers I) fun _ => ℕ → A → A := ⟨fun hI => hI.dpow⟩
+
+/- -- The synTaut linter complains about this theorem.
+theorem coe_apply {A : Type*} [CommSemiring A]
+    (I : Ideal A) (hI : DividedPowers I) (n : ℕ) (a : A) :
+  hI n a = hI.dpow n a := rfl -/
+
+theorem coe_injective {A : Type*} [CommSemiring A] (I : Ideal A) :
+    Function.Injective (fun (h : DividedPowers I) ↦ (h : ℕ → A → A)) := fun hI hI' h ↦ by
+  ext n x
+  exact congr_fun (congr_fun h n) x
+
+@[ext]
+theorem eq_of_eq_on_ideal {A : Type*} [CommSemiring A]
+    {I : Ideal A} (hI : DividedPowers I) (hI' : DividedPowers I)
+    (h_eq : ∀ (n : ℕ) {x : A} (_ : x ∈ I), hI.dpow n x = hI'.dpow n x) :
+    hI = hI' := by
+  ext n x
+  by_cases hx : x ∈ I
+  · exact h_eq n hx
+  · rw [hI.dpow_null hx, hI'.dpow_null hx]
```

Comment:
It would be nice to have this lemma as `DividedPowers.ext`, which requires removing the `ext` attribute on the structure above, and some adjustements here:
```suggestion
@[ext]
theorem DividedPowers.ext {A : Type*} [CommSemiring A]
    {I : Ideal A} (hI : DividedPowers I) (hI' : DividedPowers I)
    (h_eq : ∀ (n : ℕ) {x : A} (_ : x ∈ I), hI.dpow n x = hI'.dpow n x) :
    hI = hI' := by
  obtain ⟨hI, h₀, _⟩ := hI
  obtain ⟨hI', h₀', _⟩ := hI'
  simp only [mk.injEq]
  ext n x
  by_cases hx : x ∈ I
  · exact h_eq n hx
  · rw [h₀ hx, h₀' hx]
```
Then, `coe_injective` should also be moved after this lemma.